File: code/python/09-skills/skilled_agent.py
"""SkilledAgent — an agent that loads capabilities from a SkillRegistry.

Instead of raw tools the agent uses Skills. Each Skill provides:
  - Its own OpenAI function-calling schema (no manual tool definitions)
  - Its own prompt fragment (injected into the system prompt)
  - Validation, normalisation, and fallback (transparent to the agent loop)

The agent code does not change when skills are added or removed.

See: docs/02-the-agent-loop/05-skills-composing-capabilities.md
"""

from __future__ import annotations

import json
import logging
import os
from dataclasses import dataclass, field
from typing import Optional

from openai import OpenAI

from skill_base import Skill, SkillRegistry, SkillResult

logger = logging.getLogger(__name__)

MAX_ITERATIONS = 10

_BASE_PROMPT = """\
You are a helpful assistant with access to skills.
Answer questions using the skills provided. Always use a skill when the
question falls within its scope — never guess data that a skill can provide.
"""


# ---------------------------------------------------------------------------
# Result type
# ---------------------------------------------------------------------------


@dataclass
class SkillAgentResult:
    """The result of a single SkilledAgent.run() call."""

    answer: str
    tool_calls: list[dict] = field(default_factory=list)
    skill_results: list[dict] = field(default_factory=list)


# ---------------------------------------------------------------------------
# Agent
# ---------------------------------------------------------------------------


class SkilledAgent:
    """An agent that uses Skills instead of raw tools.

    Skills provide their own prompt fragments, validation, normalisation,
    and fallback behaviour — the agent loop doesn't need to know any of that.
    """

    def __init__(
        self,
        registry: SkillRegistry,
        model: str = "gpt-4o",
        client: Optional[object] = None,
    ) -> None:
        self.registry = registry
        self.model = model
        # Accept an injected client to support testing without API keys
        self.client: OpenAI = client or OpenAI(  # type: ignore[assignment]
            api_key=os.environ.get("OPENAI_API_KEY")
        )
        self.loaded_skills: list[Skill] = []

    # ------------------------------------------------------------------
    # Skill loading
    # ------------------------------------------------------------------

    def load_skills(self, skill_names: list[str]) -> None:
        """Load skills by name, resolving dependencies automatically.

        Each skill's dependencies are loaded before the skill itself.
        Already-loaded skills are skipped to avoid duplicates.
        """
        loaded_set: set[str] = {s.name for s in self.loaded_skills}
        for name in skill_names:
            skill = self.registry.get(name)
            for dep_skill in self.registry.resolve_dependencies(skill):
                if dep_skill.name not in loaded_set:
                    self.loaded_skills.append(dep_skill)
                    loaded_set.add(dep_skill.name)
        logger.info("Loaded skills: %s", [s.name for s in self.loaded_skills])

    # ------------------------------------------------------------------
    # Prompt construction
    # ------------------------------------------------------------------

    def build_system_prompt(self) -> str:
        """Build system prompt from base prompt + all loaded skill fragments."""
        if not self.loaded_skills:
            return _BASE_PROMPT

        fragments = []
        for skill in self.loaded_skills:
            frag = skill.get_prompt_fragment()
            fragments.append(f"## Skill: {skill.name}\n{frag}")

        return _BASE_PROMPT + "\n\n" + "\n\n".join(fragments)

    # ------------------------------------------------------------------
    # Agent loop
    # ------------------------------------------------------------------

    def run(self, user_input: str) -> SkillAgentResult:
        """Run the ReAct loop using skill schemas as tools.

        The agent reasons, calls skills (which run through their full
        validate → execute → normalise pipeline), observes the results,
        and loops until it produces a final text answer.

        Args:
            user_input: The user's question or instruction.

        Returns:
            SkillAgentResult with the final answer and execution details.

        Raises:
            ValueError: If user_input is empty.
        """
        if not user_input.strip():
            raise ValueError("user_input must not be empty")

        tools = [s.get_openai_schema() for s in self.loaded_skills]
        skill_map = {s.name: s for s in self.loaded_skills}

        messages: list[dict] = [
            {"role": "system", "content": self.build_system_prompt()},
            {"role": "user", "content": user_input},
        ]

        all_tool_calls: list[dict] = []
        all_skill_results: list[dict] = []

        for iteration in range(MAX_ITERATIONS):
            logger.debug("Iteration %d", iteration)

            response = self.client.chat.completions.create(
                model=self.model,
                messages=messages,
                tools=tools if tools else None,
            )

            msg = response.choices[0].message
            messages.append(msg)

            if not msg.tool_calls:
                return SkillAgentResult(
                    answer=msg.content or "",
                    tool_calls=all_tool_calls,
                    skill_results=all_skill_results,
                )

            for tc in msg.tool_calls:
                fn_name = tc.function.name
                try:
                    params = json.loads(tc.function.arguments)
                except json.JSONDecodeError:
                    params = {}

                logger.info("Calling skill '%s' params=%r", fn_name, params)
                all_tool_calls.append({"skill": fn_name, "params": params})

                skill = skill_map.get(fn_name)
                if skill:
                    result: SkillResult = skill.execute(params)
                else:
                    result = SkillResult(
                        success=False,
                        error=f"Unknown skill: {fn_name}",
                        error_type="internal",
                    )

                all_skill_results.append({"skill": fn_name, "result": result})

                content = (
                    json.dumps(result.data)
                    if result.success
                    else json.dumps(
                        {
                            "error": result.error,
                            "error_type": result.error_type,
                            "suggestion": result.suggestion,
                        }
                    )
                )

                messages.append(
                    {
                        "role": "tool",
                        "tool_call_id": tc.id,
                        "content": content,
                    }
                )

        logger.warning("Reached max iterations (%d)", MAX_ITERATIONS)
        return SkillAgentResult(
            answer="I wasn't able to complete the task within the iteration limit.",
            tool_calls=all_tool_calls,
            skill_results=all_skill_results,
        )


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------


def main() -> None:
    logging.basicConfig(level=logging.INFO, format="%(levelname)s %(message)s")

    from skills.weather_skill import create_weather_skill
    from skills.stock_price_skill import create_stock_price_skill
    from skills.stock_analysis_skill import create_stock_analysis_skill

    registry = SkillRegistry()
    registry.register(create_weather_skill())
    registry.register(create_stock_price_skill())
    registry.register(create_stock_analysis_skill(registry))

    agent = SkilledAgent(registry)
    agent.load_skills(["weather_reporting", "stock_analysis"])

    print("=== System Prompt ===")
    print(agent.build_system_prompt())
    print()

    queries = [
        "What's the weather in Tokyo?",
        "Should I invest in AAPL?",
        "Compare the weather in London and New York",
    ]

    for query in queries:
        print(f"=== Query: {query} ===")
        result = agent.run(query)
        print(f"Answer: {result.answer}")
        print(f"Skills used: {[tc['skill'] for tc in result.tool_calls]}")
        print()


if __name__ == "__main__":
    main()

File: code/nodejs/09-skills/skilled_agent.ts
/**
 * SkilledAgent — an agent that loads capabilities from a SkillRegistry.
 *
 * Instead of raw tools the agent uses Skills. Each Skill provides:
 *   - Its own OpenAI function-calling schema
 *   - Its own prompt fragment (injected into the system prompt)
 *   - Validation, normalisation, and fallback
 *
 * See: docs/02-the-agent-loop/05-skills-composing-capabilities.md
 */

import OpenAI from "openai";
import { Skill, SkillRegistry, SkillResult } from "./skill_base.js";

const MAX_ITERATIONS = 10;

const BASE_PROMPT = `You are a helpful assistant with access to skills.
Answer questions using the skills provided. Always use a skill when the
question falls within its scope — never guess data that a skill can provide.`;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface SkillAgentResult {
  answer: string;
  toolCalls: Array<{ name: string; args: Record<string, unknown> }>;
  skillResults: Array<{ skill: string; result: SkillResult }>;
  iterations: number;
}

// ---------------------------------------------------------------------------
// SkilledAgent
// ---------------------------------------------------------------------------

export class SkilledAgent {
  private loadedSkills: Skill[] = [];
  private client: OpenAI;

  constructor(
    private registry: SkillRegistry,
    private model = "gpt-4o",
    client?: OpenAI
  ) {
    this.client = client ?? new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
  }

  /** Load skills by name, resolving dependencies automatically. */
  loadSkills(skillNames: string[]): void {
    const loadedSet = new Set(this.loadedSkills.map((s) => s.name));
    for (const name of skillNames) {
      const skill = this.registry.get(name);
      if (!skill) continue;
      // Resolve dependencies first
      for (const dep of this.registry.resolveDependencies(skill)) {
        if (!loadedSet.has(dep.name)) {
          this.loadedSkills.push(dep);
          loadedSet.add(dep.name);
        }
      }
    }
  }

  /** Build system prompt with loaded skill fragments. */
  buildSystemPrompt(): string {
    if (!this.loadedSkills.length) return BASE_PROMPT;
    const fragments = this.loadedSkills.map(
      (s) => `## Skill: ${s.name}\n${s.getPromptFragment()}`
    );
    return BASE_PROMPT + "\n\n" + fragments.join("\n\n");
  }

  /** Run the agent loop with skill-as-tools. */
  async run(userInput: string): Promise<SkillAgentResult> {
    if (!userInput.trim()) throw new Error("userInput must not be empty");

    const tools = this.loadedSkills.map((s) => s.getOpenAISchema());
    const skillMap = new Map(this.loadedSkills.map((s) => [s.name, s]));

    const messages: OpenAI.ChatCompletionMessageParam[] = [
      { role: "system", content: this.buildSystemPrompt() },
      { role: "user", content: userInput },
    ];

    const allToolCalls: SkillAgentResult["toolCalls"] = [];
    const allSkillResults: SkillAgentResult["skillResults"] = [];

    for (let iteration = 0; iteration < MAX_ITERATIONS; iteration++) {
      const response = await this.client.chat.completions.create({
        model: this.model,
        messages,
        tools: tools.length ? tools : undefined,
        tool_choice: tools.length ? "auto" : undefined,
      });

      const choice = response.choices[0];
      const msg = choice.message;
      messages.push(msg as OpenAI.ChatCompletionMessageParam);

      if (choice.finish_reason === "stop" || !msg.tool_calls?.length) {
        return {
          answer: msg.content ?? "",
          toolCalls: allToolCalls,
          skillResults: allSkillResults,
          iterations: iteration + 1,
        };
      }

      // Execute tool calls
      for (const tc of msg.tool_calls) {
        const skillName = tc.function.name;
        let args: Record<string, unknown> = {};
        try { args = JSON.parse(tc.function.arguments); } catch { /* ignore */ }

        allToolCalls.push({ name: skillName, args });

        const skill = skillMap.get(skillName);
        let resultStr: string;
        if (skill) {
          const result = await skill.execute(args);
          allSkillResults.push({ skill: skillName, result });
          resultStr = result.success
            ? JSON.stringify(result.data)
            : `Error: ${result.error}`;
        } else {
          resultStr = `Error: Unknown skill "${skillName}"`;
        }

        messages.push({
          role: "tool",
          tool_call_id: tc.id,
          content: resultStr,
        });
      }
    }

    return {
      answer: "Maximum iterations reached without a final answer.",
      toolCalls: allToolCalls,
      skillResults: allSkillResults,
      iterations: MAX_ITERATIONS,
    };
  }
}
